### PR TITLE
Improvements: skip malformed files; use all available cores.

### PR DESCRIPTION
**Pull request by seansfkelley**
This PR introduces two changes:

### Ignoring Malformed/Empty Files

I use Backblaze as my cloud storage provider, and as an implementation detail it (or Cloud Sync, unsure) scatters empty `.bzEmpty` files to represent "folders", since AFAIK it's actually just a key-value store.

Trying to decrypt these files causes the tool to die while parsing the header. Now, instead, it will skip files with missing or malformed headers.

I don't know a good way to surface this back in the UI; I was tailing the log file to watch what it was doing and where it was failing.

### Use All Available Cores

I used the standard library `multiprocessing.Pool` to multi-process the decryption process at a per-file level. This required some rearrangement of the way that the directory hierarchy was traversed, but I think this whole program operates under the assumption that the source files/folders will not be written to while decryption is happening, so it should be okay.

This solidly pegs all 8 of my late 2018 MBP's cores, which is what I wanted.

### CODE REVIEW

File: Synology.py [hunk 1]
```diff
@@ -198,7 +198,7 @@ def validate():
                     try:
                         run_tool()
                     except Exception as e:
-                        elogger.error(e)
+                        elogger.error(e, exc_info=e)
```
Comment (seansfkelley) [inline Synology.py:201]: This change will have it print a stack trace to the log file rather than just the error message.

File: .gitignore [hunk 2]
```diff
@@ -0,0 +1,2 @@
+__pycache__
+.DS_Store
```
Comment (anojht) [inline .gitignore]: I have a gitignore file setup to exclude much more items from my local Mac as well as the gitignore file itself. I don't want to merge this change because I really think people should have their own gitignore setup, which should ignore their gitignore file itself.

File: Synology.py (hunk 1 above)
Comment (anojht) [reply Synology.py:201]: This is a needed change, thanks for this.

File: syndecrypt/__main__.py [hunk 3]
```diff
@@ -60,22 +62,36 @@ def main(args):
...
+                    os.path.join(input_dir, filename),
+                    os.path.join(input_dir.replace(ff, output_dir, 1), filename),
+                    password,
+                    private_key,
+                    public_key,
+                ))
+
+        with Pool() as p:
+            p.starmap(files.decrypt_file, decrypt_args)
+
```
Comment (anojht) [inline syndecrypt/__main__.py:94]: I remember the recursive directory traversal code took me a long time because there were a lot of cases/directory hierarchies I had to test. 

Can you ensure this remains functional with different folder/file hierarchies? Please test for the output structure as well to ensure that decryption recreates the file hierarchy properly.

**Review by anojht — COMMENTED** [review]
Thanks for this PR, you have contributed a lot of useful quality changes. Please see the comments I have left. 

Also, while I will accept this PR after the changes...In the future please open separate PRs for each new feature. So that we can keep changes clean and easily revert problematic changes.

File: .gitignore (hunk 2 above)
Comment (seansfkelley) [reply .gitignore]: Can't say I've ever heard of that gitignore philosophy, but it's your project. :) Done!

File: syndecrypt/__main__.py (hunk 3 above)
Comment (seansfkelley) [reply syndecrypt/__main__.py:94]: I've been testing these changes out with a `.sparsebundle` Time Machine backup which is a mix of files and folders and it works just the same as it did.

The code you originally had would make directories as it decrypted the files that needed. These changes split the loop so that it's two pass: first to create the directory hierarchy in serial then second to populate it with all the files in parallel. I did it this way so that the main decryption loop was simpler to parallelize given the signature of `decrypt_file`, which does not create any necessary directories.

So in short, the logic hasn't changed, it's just reordered.

As noted in the PR description, this implementation assumes that the source directory isn't being written to and that the destination directory is only being written to by this process. If anyone else interferes while it's running, all bets are off, but that was already true.

File: .gitignore (hunk 2 above)
Comment (anojht) [reply .gitignore]: Thanks, this is just because everyone has a different dev environment, that may create different sets of temporary files. So a single gitignore file isn't going to catch all cases, that's why I believe people should setup their environments to their liking and learn how to manage their gitignore file. 😄

**Comment by seansfkelley** [comment]
> In the future please open separate PRs for each new feature. So that we can keep changes clean and easily revert problematic changes.

Yeah, I can sympathize with that; the parallelization was kind of tacked on the end and it would have merge conflicted with the other changes so I grouped it together.

Thanks for doing the legwork to reverse engineer the format! I found this repo from a forum post on Synology and I hope other have found it as useful as I have (and I wanted to contribute back).

File: syndecrypt/__main__.py (hunk 3 above)
Comment (anojht) [reply syndecrypt/__main__.py:94]: Yes I see that you store the list of directories, make one pass to recreate the structure then another pass to create a list of files to be decrypted and execute the decryption call in parallel. Very nice work!

**Comment by anojht** [comment]
Thanks, I couldn't believe Synology for all these years have yet to release a tool for one of the most popular platforms out there. Luckily someone had a really well documented repo, although outdated and nonfunctional, for me to work from. Saved a lot of time and Python made it quick 😄 

I originally wanted to write the app in Swift and publish it, but decided this would make it easier to maintain and make it easily portable to other platforms. Already this decision is paying off! You have contributed tremendously to improve the tool, so thanks for donating your time!

As for the PR, everything looks good and I am merging this. I will publish a new release tomorrow!

**Comment by seansfkelley** [comment]
Awesome, thanks! I am also surprised they haven't bothered to release a Mac-compatible tool, but I would much rather use an open-source solution like this one than rely on mystery closed-source program anyway.
